server/src/main/java/org/apache/druid/segment/loading/RoundRobinStorageLocationSelectorStrategy.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.segment.loading;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterators;
+import org.apache.druid.timeline.DataSegment;
+
+import java.util.Iterator;
+
+/**
+ * A {@link StorageLocation} selector strategy that selects a segment cache location in a round-robin fashion each time
+ * among the available storage locations.
+ */
+public class RoundRobinStorageLocationSelectorStrategy implements StorageLocationSelectorStrategy
+{
+  private Iterator<StorageLocation> cyclicIterator;
+
+  @Override
+  public void setStorageLocations(ImmutableList<StorageLocation> storageLocations)
+  {
+    // cyclicIterator remembers the marker internally
+    cyclicIterator = Iterators.cycle(storageLocations);
+  }
+
+  @Override
+  public Iterator<StorageLocation> getLocations(DataSegment dataSegment, String storageDirStr)
+  {
+    return cyclicIterator;
```

Comment:
This iterator should count the number of iterations and `hasNext()` should return false when all locations in the iterator is iterated.

Reply:
Looking at this.

Reply:
without that this is buggy

Reply:
Trying to implement a custom cyclic iterator which keeps track of the no. of iterations (`numLocationsToTry`) . The implementation seems slightly tricky to me. Similar to the `leastBytes` used selector, synchronization issues can occur with `roundRobin` iterator when multiple threads are loading segments. This might lead to an incorrect round robin behavior. Please suggest a clean way to handle this.

Reply:
> The implementation seems slightly tricky to me. Similar to the leastBytes used selector, synchronization issues can occur with roundRobin iterator when multiple threads are loading segments. This might lead to an incorrect round robin behavior. 

Good point. An example implementation could be:

```java
public class RoundRobinStorageLocationSelectorStrategy implements StorageLocationSelectorStrategy
{
  private final List<StorageLocation> storageLocations;
  private final AtomicInteger startIndex = new AtomicInteger(0);

  public RoundRobinStorageLocationSelectorStrategy(List<StorageLocation> storageLocations)
  {
    this.storageLocations = storageLocations;
  }

  @Override
  public Iterator<StorageLocation> getLocations()
  {
    return new Iterator<StorageLocation>()
    {
      private final int numStorageLocations = storageLocations.size();
      private int remainingIterations = numStorageLocations;
      private int i = startIndex.getAndIncrement();

      @Override
      public boolean hasNext()
      {
        return remainingIterations > 0;
      }

      @Override
      public StorageLocation next()
      {
        if (!hasNext()) {
          throw new NoSuchElementException();
        }
        remainingIterations--;
        final StorageLocation nextLocation = storageLocations.get(i++);
        if (i == numStorageLocations) {
          i = 0;
        }
        return nextLocation;
      }
    };
  }
}
```

Reply:
Thanks for the example code. 
1. `startIndex` goes out of bounds. So used `AtomicLong`'s `getAndUpdate()` to do an increment with a modulo operation.
`private int i = startIndex.getAndUpdate(n -> (n + 1) % numStorageLocations);`
2. `startIndex` had to be reset when all the locations are iterated.
```
        if (i == numStorageLocations) {
          i = 0;
          startIndex.set(0);
        }
```
Tested this cyclic iterator with StorageLocationSelectorStrategyTest and few other tests.

Reply:
Oh right! Thanks for catching it.

Reply:
One question is, why is `2.` required? If I understand correctly, the first storage location is picked in the round robin manner. And the returned iterator from `getLocations()` just iterates all remaining storage locations. Not sure why this iterator should reset `startIndex`.

Reply:
I've a test case (single thread sequential calls to the iterator) where the round robin iteration becomes incorrect with or without the fix in point 2. The fix I think is to use a single variable alone `startIndex`, `i` may not be needed. A dry run of the test case testRoundRobinLocationSelectorStrategy() with calls to next() gives below indices. The locations are not being iterated in proper round robin fashion.

> i=0, startIndex=1
> i=1, startIndex=1
> i=2, startIndex=1
> i=1, startIndex=2
> i=2, startIndex=2
> i=0, startIndex=2
> i=2, startIndex=0
> i=0, startIndex=0
> i=1, startIndex=0
> i=0, startIndex=1
> i=1, startIndex=1
> i=2, startIndex=1
> i=1, startIndex=2
> i=2, startIndex=2
> i=0, startIndex=2

The code will look like,
```
public class RoundRobinStorageLocationSelectorStrategy implements StorageLocationSelectorStrategy
{

  private final List<StorageLocation> storageLocations;
  private final AtomicInteger startIndex = new AtomicInteger(0);

  public RoundRobinStorageLocationSelectorStrategy(List<StorageLocation> storageLocations)
  {
    this.storageLocations = storageLocations;
  }

  @Override
  public Iterator<StorageLocation> getLocations()
  {
    return new Iterator<StorageLocation>() {

      private final int numStorageLocations = storageLocations.size();
      private int remainingIterations = numStorageLocations;

      @Override
      public boolean hasNext()
      {
        return remainingIterations > 0;
      }

      @Override
      public StorageLocation next()
      {
        if (!hasNext()) {
          throw new NoSuchElementException();
        }
        remainingIterations--;
        final StorageLocation nextLocation =
          storageLocations.get(startIndex.getAndUpdate(n -> (n + 1) % numStorageLocations));
        return nextLocation;
      }
    };
  }

}
```
This should likely resolve the issues. I'm further testing this.
```
final StorageLocation nextLocation =
          storageLocations.get(startIndex.getAndUpdate(n -> (n + 1) % numStorageLocations));
```

Reply:
Fixed.

Reply:
Oh, now I know what the round robin you want is. What I thought was, each caller will get an iterator with a different `startIndex` which is changed in a round robin fashion. Okay, your implementation makes sense. Please add more details description about the behavior of this strategy especially when multiple threads use this.